lib/stdio/pipeline.js

```diff
@@ -12,56 +13,48 @@ export const pipeStreams = (sources, destination) => {
 	} else {
 		const mergedSource = mergeStreams(sources);
 		mergedSource.pipe(destination);
-		handleDestinationComplete(mergedSource, destination, finishedStreams);
+		handleStreamError(destination, mergedSource, finishedStreams);
 	}
 
 	for (const source of sources) {
-		handleSourceAbortOrError(source, destination, finishedStreams);
-		handleDestinationComplete(source, destination, finishedStreams);
+		handleStreamError(source, destination, finishedStreams);
+		handleStreamError(destination, source, finishedStreams);
 	}
 };
 
 // `source.pipe(destination)` makes `destination` end when `source` ends.
 // But it does not propagate aborts or errors. This function does it.
-const handleSourceAbortOrError = async (source, destination, finishedStreams) => {
-	if (isStandardStream(source) || isStandardStream(destination)) {
+// We do the same thing in the other direction as well.
+const handleStreamError = async (stream, otherStream, finishedStreams) => {
+	if (isStandardStream(stream) || isStandardStream(otherStream)) {
 		return;
 	}
 
 	try {
-		await onFinishedStream(source, finishedStreams);
+		await onFinishedStream(stream, finishedStreams);
 	} catch (error) {
-		destroyStream(destination, finishedStreams, error);
-	}
-};
-
-// The `destination` should never complete before the `source`.
-// If it does, this indicates something abnormal, so we abort or error `source`.
-const handleDestinationComplete = async (source, destination, finishedStreams) => {
-	if (isStandardStream(source) || isStandardStream(destination)) {
-		return;
-	}
-
-	try {
-		await onFinishedStream(destination, finishedStreams);
-		destroyStream(source, finishedStreams);
-	} catch (error) {
-		destroyStream(source, finishedStreams, error);
```

Comment:
The changes in this file are copied from ongoing PR #821. And also the other changes in that PR, such as the new tests.